cmake: Remove absolute paths and private links from exported targets.

ryanvolz opened the pull request:
This transitions to using `Python::NumPy`, `pybind11::pybind11`, and `libunwind::libunwind` targets for building with the include directories/libraries of NumPy, pybind11, and libunwind. The Python::NumPy target matches what is available with the FindPython module in CMake 3.14+ (and can be transitioned to such when the minimum CMake version is past that). These changes have the effect of cleaning up the include_directory paths added to the gnuradio-runtime target so that no absolute paths are used, which is helpful for relocatable installations (e.g. conda).

Part of this change involves moving some link targets to PRIVATE since they are not actually part of the public interface. In the case of the python bindings, these are not exported and have no public interface for consumption by other C++ libraries, so the switch from PUBLIC to PRIVATE has no practical effect but could still help avoid future confusion.

Fixes #4380.

noc0lour commented on cmake/Modules/GnuradioConfig.cmake.in:
```diff
@@ -36,6 +36,8 @@ if(${ENABLE_PYTHON})
   include(GrPython)
 endif()
+# pybind11 needed even without Python enabled for building against gnuradio-runtime
+find_dependency(pybind11)
```
Hm, so it is not possible to build GNU Radio runtime withouth pybind11/Python? Not that I mind, but it's a bit surprising for the C++ interface to still require pybind11.

ryanvolz replied:
I can't say I understand how this all works, but I think that it's a required dependency for `gnuradio-runtime` and thus all OOT modules because of the header inclusion in `block_gateway.h` [here](https://github.com/gnuradio/gnuradio/blob/1a0be2e6b54496a8136a64d86e372ab219c6559b/gnuradio-runtime/include/gnuradio/block_gateway.h#L18). I believe that is also the reason that it is necessary to build GNU Radio and any OOTs with the **same** version (or a compatible version) of pybind11, as has recently come up a few times on the mailing list. (Maybe there is something we should be doing better there with this CMake Config file to prevent bad builds). I haven't actually looked through the header or the code to see what exactly it needs pybind11 for.

mormj replied:
I like these changes to better export the targets, but we need to figure out why/if pybind11 is/has always been required for non-python builds and fix that too.

ryanvolz replied:
The issue looks to be only `block_gateway`, which currently needs `pybind11` but it's also built into `gnuradio-runtime` regardless of if Python support is enabled. The commit where `pybind11` was introduced is 0249f7ce0cf5173b946b936c5cd66380afc8bf92, so thankfully we have someone here who should know what's going on there 😁. I haven't looked at this code to try to understand it, but I imagine `block_gateway` *could* be put behind `ENABLE_PYTHON`? Is there any way it could work or be rewritten for someone to build the runtime without `pybind11` but then still give modules (including runtime itself, I suppose) the option of building Python bindings?

mormj replied:
That is a quandary - similar setup we had had with block_gateway previously being dependent on swig director functionality - I'll have to go back and look at 3.8 to see how we were able to build gr without python/swig.  Maybe it just works now with pybind11 and no python - i just haven't tried it yet. Since pybind11 is just headers

ryanvolz replied:
I suspect that it would work fine as-is now with just `pybind11` and no `ENABLE_PYTHON`. The caveat with needing `pybind11` in this manner is that it *does* introduce a version constraint on the `pybind11` that is later used to build the bindings. We've seen reports of this. AFAIK it doesn't have to be the exact same version, but only ones with the same `PYBIND11_INTERNALS_VERSION`. So it might be nice on a couple fronts to eliminate `block_gateway`'s dependence on `pybind11`, if it's possible.

noc0lour commented on gnuradio-runtime/lib/CMakeLists.txt:
```diff
@@ -131,23 +128,22 @@ target_link_libraries(gnuradio-runtime PUBLIC
   Log4Cpp::log4cpp
   MPLib::mplib
-  ${libunwind_LIBRARIES}
+  pybind11::pybind11
```
Or is pybind11 required because of this (new) statement?

ryanvolz replied:
This `pybind11::pybind11` target only adds the include path for the pybind11 headers (and the Python headers, but we already do that anyway). So effectively it replaces the now-removed `target_include_directories` statement, and it maintains the functionality of communicating to downstream consumers of the `gnuradio-runtime` CMake target where they can find the pybind11 headers needed for `block_gateway.h`.

noc0lour reviewed (changes requested):
I'm not 100% sure if the mandatory pybind11 dependency is only due to inclusion in `CMakeLists.txt` or because of other dependencies in the source code.  Everything else looks fine since it affects only the python module and this module correctly doesn't need to expose a public interface even on CMake side.

mormj commented:
@noc0lour - do the responses to your concerns seem sufficient enough to go ahead and merge this, or are you still concerned?

ryanvolz commented:
I've updated this based on building Gqrx against GNU Radio with the previous patch and thus identifying a case where pybind11 is definitely not needed and should not always be required through `GnuradioConfig.cmake`. Therefore I've removed the `find_dependency(pybind11)` from `GnuradioConfig.cmake.in` and moved `pybind11::pybind11` from a `PUBLIC` target dependency of `gnuradio-runtime` to `PRIVATE`. I think these changes should satisfy the questions raised so far.

Here's the state of pybind11 as I currently understand it:

- pybind11 is always required to build gnuradio-runtime, because it is used in `block_gateway`, regardless of `ENABLE_PYTHON`.
- pybind11 is *not* always required to build external projects/OOTs against gnuradio-runtime, unless they were to consume the `block_gateway.h` header (**Does this actually need to be a public header?**).
- The pybind11 used to build an OOT must have the same `PYBIND11_INTERNALS_VERSION` as the pybind11 used to build gnuradio-runtime (evidenced by user reports of build errors), but I don't know if a restructuring could fix this. Conversely, I don't see a great way of ensuring a compatible `pybind11` through CMake short of requiring the exact same version, which seems too strict.

mormj commented:
Let's go ahead and merge this, and deal with the block_gateway dependence on pybind later.  Pybind is a very light dependency, but should not be a requirement for a pythonless build - but we can take that up separately

willcode commented:
Opinions on backport? Any ripple effect wrt deps?

ryanvolz commented:
> Opinions on backport? Any ripple effect wrt deps?

I've used this in the 3.9 conda package and have not seen any issues with gr-osmosdr or gr-satellites, so I expect it's safe enough and worth a backport to 3.9 to help the fix for #4380 propagate further and sooner.
